pkg/operator/configobservation/cloudprovider/observe_cloudprovider.go

```diff
@@ -0,0 +1,130 @@
+package cloudprovider
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+)
+
+const (
+	targetNamespaceName       = "openshift-kube-apiserver"
+	cloudProviderConfFilePath = "/etc/kubernetes/static-pod-resources/configmaps/cloud-config/%s"
+	configNamespace           = "openshift-config"
+)
+
+func ObserveCloudProviderNames(
+	genericListers configobserver.Listers,
+	recorder events.Recorder,
+	existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	var errs []error
+	cloudProvidersPath := []string{"extendedArguments", "cloud-provider"}
+	cloudProviderConfPath := []string{"extendedArguments", "cloud-config"}
+	previouslyObservedConfig := map[string]interface{}{}
+
+	existinCloudConfig, _, err := unstructured.NestedStringSlice(existingConfig, cloudProviderConfPath...)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	if currentCloudProvider, _, _ := unstructured.NestedStringSlice(existingConfig, cloudProvidersPath...); len(currentCloudProvider) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentCloudProvider, cloudProvidersPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(existinCloudConfig) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, existinCloudConfig, cloudProviderConfPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	observedConfig := map[string]interface{}{}
+
+	infrastructure, err := listers.InfrastructureLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserverCloudProviderNames", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		return observedConfig, errs
+	}
+	if err != nil {
+		return previouslyObservedConfig, errs
+	}
+
+	cloudProvider := getPlatformName(infrastructure.Status.Platform, recorder)
```

Comment:
i would not pass recorder here, but check the error conditions before this function.